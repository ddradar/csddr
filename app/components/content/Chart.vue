<template>
  <UBadge
    v-for="(level, i) in levels"
    :key="i"
    :title="colors[charts[i]].name"
    :color="colors[charts[i]].color"
    v-bind="rest"
  >
    {{ level }}
  </UBadge>
</template>

<script lang="ts" setup>
import type { BadgeProps } from '#ui/types'

interface ChartProps extends /* @vue-ignore */ Omit<BadgeProps, 'color'> {
  colors?: Record<number, { name: string; color: BadgeProps['color'] }>
  levels: (number | '?' | '10+')[]
  charts?: number[]
  dp?: boolean
}

const props = withDefaults(defineProps<ChartProps>(), {
  colors: () => ({
    0: { name: 'BEGINNER', color: 'info' },
    1: { name: 'BASIC', color: 'warning' },
    2: { name: 'DIFFICULT', color: 'error' },
    3: { name: 'EXPERT', color: 'primary' },
    4: { name: 'CHALLENGE', color: 'secondary' },
  }),
  charts: p =>
    p.colors
      ? Object.entries(p.colors)
          .filter(([_, d]) => !!d)
          .map(([i]) => parseInt(i, 10))
      : p.dp
        ? [0, 1, 2, 3, 4]
        : [1, 2, 3, 4],
})
const rest = computed(() => {
  const { colors: _1, levels: _2, charts: _3, dp: _4, ...rest } = props
  return rest
})
</script>
