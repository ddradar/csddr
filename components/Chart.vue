<template>
  <Badge
    v-for="(level, i) in levels"
    :key="i"
    :title="types[charts[i]].name"
    :type="types[charts[i]].type"
  >
    {{ level }}
  </Badge>
</template>

<script lang="ts" setup>
interface ChartProps {
  types?: Record<number, { name: string; type: string }>
  levels: (number | '?' | '10+')[]
  charts?: number[]
}
const _defaultTypes = {
  0: { name: 'BEGINNER', type: 'info' },
  1: { name: 'BASIC', type: 'warning' },
  2: { name: 'DIFFICULT', type: 'danger' },
  3: { name: 'EXPERT', type: 'success' },
  4: { name: 'CHALLENGE', type: 'primary' },
}

withDefaults(defineProps<ChartProps>(), {
  types: () => _defaultTypes,
  charts: p =>
    Object.keys(p.types ?? _defaultTypes).map(s => parseInt(s, 10)),
})
</script>
