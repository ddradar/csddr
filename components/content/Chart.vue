<template>
  <Badge
    v-for="(level, i) in levels"
    :key="i"
    :title="types[charts[i]].name"
    :type="types[charts[i]].type"
  >
    {{ level }}
  </Badge>
</template>

<script lang="ts" setup>
interface ChartProps {
  types?: Record<number, { name: string; type: string }>
  levels: (number | '?' | '10+')[]
  charts?: number[]
}

withDefaults(defineProps<ChartProps>(), {
  types: () => ({
    0: { name: 'BEGINNER', type: 'info' },
    1: { name: 'BASIC', type: 'warning' },
    2: { name: 'DIFFICULT', type: 'danger' },
    3: { name: 'EXPERT', type: 'success' },
    4: { name: 'CHALLENGE', type: 'primary' },
  }),
  charts: p =>
    Object.keys(p.types ?? { 0: '', 1: '', 2: '', 3: '', 4: '' }).map(s =>
      parseInt(s, 10)
    ),
})
</script>
