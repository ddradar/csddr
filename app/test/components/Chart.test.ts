import { describe, expect, test } from 'vitest'
import Chart from '~/components/content/Chart.vue'
import { mount } from '@vue/test-utils'

describe('/app/components/content/Chart.vue', () => {
  const colors = [
    { name: 'NORMAL', color: 'yellow' },
    { name: 'BASIC', color: 'yellow' },
    { name: 'ANOTHER', color: 'red' },
    { name: 'MANIAC', color: 'green' },
  ] as const
  test.each([
    { levels: [1, 2, 3, 4, 5] },
    { levels: ['?' as const, 5.5, 13], charts: [1, 2, 3] },
    { levels: [3, 4, 5, 6], colors },
    { levels: [3, 4, 5], colors, charts: [1, 2, 3] },
  ])('%o renders correctly', (props: InstanceType<typeof Chart>['$props']) => {
    const sut = mount(Chart, { props })
    expect(sut.html()).toMatchSnapshot()
  })
})
