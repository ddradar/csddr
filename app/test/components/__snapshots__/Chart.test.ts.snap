// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/app/components/content/Chart.vue > { levels: [ '?', 5.5, 13 ], charts: [ 1, 2, 3 ] } renders correctly 1`] = `
"<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-yellow-500 dark:bg-yellow-400 text-white dark:text-gray-900" title="BASIC">?</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-red-500 dark:bg-red-400 text-white dark:text-gray-900" title="DIFFICULT">5.5</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-green-500 dark:bg-green-400 text-white dark:text-gray-900" title="EXPERT">13</span>"
`;

exports[`/app/components/content/Chart.vue > { levels: [ 1, 2, 3, 4, 5 ] } renders correctly 1`] = `
"<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-blue-500 dark:bg-blue-400 text-white dark:text-gray-900" title="BEGINNER">1</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-yellow-500 dark:bg-yellow-400 text-white dark:text-gray-900" title="BASIC">2</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-red-500 dark:bg-red-400 text-white dark:text-gray-900" title="DIFFICULT">3</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-green-500 dark:bg-green-400 text-white dark:text-gray-900" title="EXPERT">4</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-purple-500 dark:bg-purple-400 text-white dark:text-gray-900" title="CHALLENGE">5</span>"
`;

exports[`/app/components/content/Chart.vue > { levels: [ 3, 4, 5 ], colors: [ { name: 'NORMAL', color: 'yellow' }, { name: 'BASIC', color: 'yellow' }, { name: 'ANOTHER', color: 'red' }, { name: 'MANIAC', color: 'green' } ], charts: [ 1, 2, 3 ] } renders correctly 1`] = `
"<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-yellow-500 dark:bg-yellow-400 text-white dark:text-gray-900" title="BASIC">3</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-red-500 dark:bg-red-400 text-white dark:text-gray-900" title="ANOTHER">4</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-green-500 dark:bg-green-400 text-white dark:text-gray-900" title="MANIAC">5</span>"
`;

exports[`/app/components/content/Chart.vue > { levels: [ 3, 4, 5, 6 ], colors: [ { name: 'NORMAL', color: 'yellow' }, { name: 'BASIC', color: 'yellow' }, { name: 'ANOTHER', color: 'red' }, { name: 'MANIAC', color: 'green' } ] } renders correctly 1`] = `
"<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-yellow-500 dark:bg-yellow-400 text-white dark:text-gray-900" title="NORMAL">3</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-yellow-500 dark:bg-yellow-400 text-white dark:text-gray-900" title="BASIC">4</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-red-500 dark:bg-red-400 text-white dark:text-gray-900" title="ANOTHER">5</span>
<span class="inline-flex items-center font-medium rounded-md text-xs px-2 py-1 bg-green-500 dark:bg-green-400 text-white dark:text-gray-900" title="MANIAC">6</span>"
`;
